components/exercises/chapter-4-oscar/Utils.tsx

```diff
@@ -70,3 +72,139 @@ export function reverseArrayInPlace(arr: any[]): undefined {
+ */
+export function listToArray(list: ListItemProp | null): ReactNode[] {
+  let arr: ReactNode[] = [];
+
+  // node as condition because we want to stop when node is null
+  // node.rest because we want to move to the next node
+  for (let node: ListItemProp | null = list; node; node = node.rest) {
+    arr.push(node.value);
+  }
+  return arr;
+}
+
+/**
+ * Prepends a value to a list.
+ * @param value - The value to prepend to the list.
+ * @param list - The list to which the value will be prepended.
+ * @returns A new list with the value prepended to it.
+ */
+export function prepend(
+  value: number | string,
+  list: ListItemProp | null
+): ListItemProp | null {
+  // list is null when we reach the end of the list
+  return { value, rest: list };
+}
+
+/**
+ * Returns the value of the nth element in a list.
+ * @param list - The list to search for the nth element.
+ * @param n - The index of the element to retrieve.
+ * @returns The value of the nth element in the list, or undefined if the list is empty or n is out of bounds.
+ */
+export function nth(
+  list: ListItemProp | null,
+  n: number
+): undefined | ReactNode {
+  if (!list) return undefined;
+  else if (n == 0) return list.value;
+  // list.rest because we want to move to the next node
+  else return nth(list.rest, n - 1);
+}
+
+/**
+ * Retrieves the corresponding food for a given city.
+ * @param items - The list of items.
+ * @returns The food associated with the city.
+ */
+export function getFood(items: ListItemProp | null) {
+  const city = nth(items, 0);
+  let food = '';
+
+  switch (city) {
+    case 'Paris':
+      food = 'Ratatouille';
+      break;
+    case 'London':
+      food = 'Sunday roast';
+      break;
+    case 'Barcelona':
+      food = 'Paella';
+      break;
+    case 'Porto':
+      food = 'Salted cod';
+      break;
+    case 'Milan':
+      food = 'Lasagna';
+      break;
+    case 'Geneva':
+      food = 'Fondue';
+      break;
+    default:
+      food = 'Goulash';
+      break;
+  }
+  return <p>{food}</p>;
+}
+
+/**
+ * Represents a list item.
+ * @param item - The item to be rendered.
+ * @returns The rendered list item.
+ */
+export function ListItem({ item }: { item: ListItemProp }) {
+  return (
+    <li>
+      <p className='text-gray-800'>{item.value}</p>
+      {item.rest === null ? (
+        <div className='mb-2'>
+          <div className='cursor-pointer rounded-lg bg-red-600 p-3 shadow'>
+            <p className='font-semibold'>No more cities to visit</p>
+          </div>
+        </div>
+      ) : (
+        <details className='mb-2'>
+          <summary className='cursor-pointer rounded-lg bg-yellow-600 p-3 shadow'>
+            <span className='font-semibold'>
+              {item.rest && Object.getOwnPropertyNames(item?.rest)[1]}
+            </span>
+          </summary>
+          <div className='bg-white p-4'>
+            <ListItem item={item.rest} />
+          </div>
+        </details>
+      )}
+    </li>
+  );
+}
```

Comment:
Generally, functions that return JSX (i.e. components) go into their own files, or in this case I would just put ListItem in the same file as AList, since it is only used there

Reply:
understood, moved to the component now named as CityList

Reply:
Also tests moved to CityList.test.tsx